src/landscape.cpp

```diff
@@ -994,6 +994,83 @@ static bool MakeLake(TileIndex tile, void *user_data)
 	return false;
 }
 
+/**
+ * Check whether a river could (logically) flow into a lock.
+ * @param tile the middle tile of a lock.
+ * @param recursive whether the function is being called recursively.
+ * @return true iff the water can be flowing into a lock.
+ */
+static bool FlowsLock(TileIndex tile, bool recursive = false)
+{
+	DiagDirection dir = GetInclinedSlopeDirection(GetTileSlope(tile));
+	if (dir == INVALID_DIAGDIR) return false;
+
+	int delta_mid = TileOffsByDiagDir(dir);
+	if (!IsTileFlat(tile + delta_mid)) return false;
+	if (!IsTileFlat(tile - delta_mid)) return false;
+
+	if (!recursive) return true;
```

Comment:
Split into two functions. I.e. one that does the above tile checks and the main recursive function.